### PR TITLE
lib/attrsets: add cartesianProductOfSets function

### PR DESCRIPTION
**Pull request by tfc**
###### Motivation for this change

For library maintainers and other projects it is useful to build a project with multiple library and toolchain combinations.
For example, in C++ libraries it might be worthwhile to take the derivation and override it with different build profile settings and toolchains.

This might look like this:

```nix
  buildConfs = lib.cartesianProductOfSets {
    cc = with pkgs; [ clang_9 clang_10 gcc7 gcc8 gcc9 gcc10 ];
    buildType = [ "Debug" "Release" ];
  };
  builds = let
    buildFunction = { cc, buildType }: lib.nameValuePair
      "${cc.name}-${lib.toLower buildType}"
      (
        callPackage ./build.nix {
          inherit buildType;
          stdenv = overrideCC stdenv cc.value;
        }
      );
  in
    builtins.listToAttrs (builtins.map buildFunction buildConfs);
  # this results in builds = { gcc9-release = ...; gcc9-debug = ...; clang_9-release = ...; ...etc... }
```

implemented the function `cartesianProductOfSets` along with nix unit tests

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: lib/attrsets.nix [hunk 1]
```diff
@@ -494,4 +494,26 @@ rec {
   zip = builtins.trace
     "lib.zip is deprecated, use lib.zipAttrsWith instead" zipAttrsWith;
 
+  /* Return the cartesian product of attribute set value combinations.
```
Comment (blitz) [inline lib/attrsets.nix:497]: It would be extra awesome if this documentation would be in the nixpkgs manual.

Comment (tfc) [reply lib/attrsets.nix:497]: done.

File: lib/attrsets.nix [hunk 2]
```diff
@@ -494,4 +494,26 @@ rec {
...
+    crossMap = set: attrName: attrValues: let
+        f = attrValue: (set // { "${attrName}" = attrValue; });
+      in builtins.map f attrValues;
+    crossApply = attrName: attrValues: list:
+      builtins.concatMap (set: crossMap set attrName attrValues) list;
+    crossSet = set: let
+        names = builtins.attrNames set;
+        f = (list: name: crossApply name (set."${name}") list);
+      in builtins.foldl' f [{}] names;
+  in crossSet;
```
Comment (Infinisil) [inline lib/attrsets.nix:518]: I feel like the `let in` are a bit distracting, how about
```nix
{
  cartesianProductOfSets = attrsOfLists:
    lib.foldl' (listOfAttrs: attrName:
      concatMap (attrs:
        map (listValue: attrs // { ${attrName} = listValue; }) attrsOfLists.${attrName}
      ) listOfAttrs
    ) [{}] (attrNames attrsOfLists);
}
```

Instead

**Review by infinisil — COMMENTED** [review]
Note that there's also `lists.crossList` which does almost the same:
```nix
lists.crossLists (a: b: { inherit a b; }) [ [ 1 2 ] [ 10 20 ] ]
-> [ { a = 1; b = 10; } { a = 1; b = 20; } { a = 2; b = 10; } { a = 2; b = 20; } ]
```

Not sure if there's some unified way to implement these two

File: lib/attrsets.nix (hunk 2 above)
Comment (tfc) [reply lib/attrsets.nix]: you're right, looks much cleaner. i incorporated your suggestion.

**Comment by blitz** [comment]
@tfc Do you think that this function could be implemented in terms of `crossLists` or vice versa? Or do they overlap so much that its not worth having both?

**Comment by tfc** [comment]
@blitz @Infinisil i tried implementing `cartesianProductOfSets` in terms of `crossLists` and with the help of a colleague came up with:


```nix
  cartesianProductOfSets = set:
    let
      extendSetOrFunction = setOrFunction: attrName:
        if builtins.isFunction setOrFunction
        then a: extendSetOrFunction (setOrFunction a) attrName
        else a: setOrFunction // { ${attrName} = a; };
    in
    pkgs.lib.crossLists
      (pkgs.lib.foldl extendSetOrFunction { } (builtins.attrNames set))
      (builtins.attrValues set);
```

not sure how much better this is.

**Comment by Infinisil** [comment]
Tbh I think we can just deprecate `crossLists`. I think it's just an inferior version of this function here, and any usages of `crossLists` can be replaced with a more easily-readable version of this function.

**Comment by tfc** [comment]
@infinisil i deprecated `crossLists`.

**Review by blitz — APPROVED** [review]
This came out very nicely!

**Review by ghost — APPROVED** [review]
Yes please

**Comment by Infinisil** [comment]
Nice! Though the formatting makes this hard to review. Since we don't yet have a policy for auto-formatters I think we should avoid them for the time being.

**Comment by tfc** [comment]
@Infinisil shall i roll back the nixpkgs-fmt application to the code?

**Comment by Infinisil** [comment]
@tfc Yes please. I really hope we can adopt a formatter for all of nixpkgs eventually though :)

**Comment by tfc** [comment]
@Infinisil done

**Review by infinisil — APPROVED** [review]

**Review by jonringer — APPROVED** [review]
diff LGTM

**Pull request merged**